[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest

ipiszy opened the pull request:
This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108015
* #107931
* __->__ #107901
* #107847
* #107802

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ipiszy committed:
[Inductor CUTLASS backend] Step 3: autotune_process, and
CUDABenchmarkRequest.

[ghstack-poisoned]

kadeng commented on torch/_inductor/autotune_process.py:
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
 
-    input_tensors: Union["TensorMeta", List["TensorMeta"]]
-    output_tensor: Union["TensorMeta", List["TensorMeta"]]
+    def make_run_fn(
```
I think this might benefit from a small doc comment telling which kind of run function it makes

kadeng commented on torch/_inductor/autotune_process.py:
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
+    ) -> Callable[[], None]:
+        mod = PyCodeCache.load_by_key_path(self.module_cache_key, self.module_path)
+        if DEBUG:
+            print(
```
Better to use logging with a DEBUG level here.

aakhundov replied:
I guess, this approach is inherited from the existing code.

ipiszy replied:
Yes this is from original code. I refactored it a bit so that TritonBenchmarkRequest and CUDABenchmarkRequest can share code. I'm not sure whether log works correctly when benchmark is in a subprocess so don't want to just change it. Could be a follow up task.

kadeng commented on torch/_inductor/autotune_process.py:
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
+            c_void_p(tensor.data_ptr())
+            for tensor in list(input_tensors) + [output_tensor]
+        ]
+        log.debug(
```
Init logger at the top of the module first?

ipiszy replied:
Fixed.

kadeng commented on torch/_inductor/autotune_process.py:
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
+        )
+
+        # Generate partial function.
+        return functools.partial(
```
Not saying this is a bad idea here to use functools.partial, but it can make debugging harder, as functools.partial binds it into a native object, leaving no trace in callstacks where the code is coming from. So when the function returned by functools.partial is called at a different callsite later on, it is hard to figure out which function is being called, without actually using a debugger or tracing back the origin of the function. On the other hand, it's likely slightly faster than using a python wrapper function.

kadeng commented on torch/_inductor/autotune_process.py:
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
+        )
+
+    def cleanup_run_fn(self) -> None:
+        self.DLL.close()
```
Should be guarded against being closed twice I guess.

ipiszy replied:
DLLWrapper.close() already handles it.

kadeng commented on torch/_inductor/utils.py:
```diff
@@ -46,6 +46,79 @@
+    log.debug("filtered events")
+    log.debug(filtered_events)
+    res = sum(event.cuda_time for event in filtered_events) / 1000.0
+    log.debug(f"profiling results: {res}")
```
I think this might warrant logging at "INFO" level.

ipiszy replied:
Maybe end user doesn't want the benchmark logging, so still keep it as debug info.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"


CUDABenchmarkRequest.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

aakhundov commented on torch/_inductor/utils.py:
```diff
@@ -45,6 +45,79 @@
+    estimate_ms = start_event.elapsed_time(end_event) / 5
+
+    # compute number of warmup and repeat
+    n_warmup = max(1, int(warmup / estimate_ms))
+    n_repeat = max(1, int(rep / estimate_ms))
```
I'm wondering, why not use the passed `warmup` and `rep` directly, instead of computing `n_warmup` and `n_repeat` from them?

ipiszy replied:
This is copied from Triton bench. I think it makes sure that there is enough time spent to warm up the device (so that the GPU frequency is set to max). If we only pass a count, for small kernels the warm up time might not be enough, while for big kernels the warm up time could be too long. n_repeat also makes sure that the total time doesn't vary a lot for small and big kernels.

aakhundov commented on torch/_inductor/utils.py:
```diff
@@ -45,6 +45,79 @@
+    log.debug("filtered events")
+    log.debug(filtered_events)
+    res = sum(event.cuda_time for event in filtered_events) / 1000.0
+    log.debug("profiling results: %s", res)
```
Nit: add the time unit here for clarity?

aakhundov commented on torch/_inductor/autotune_process.py:
```diff
@@ -166,71 +171,197 @@ class BenchmarkRequest:
+        assert (
+            self.workspace_size == 0
+        ), "Autotune cache needs to be updated to support non-zero workspace_size!"
+        self.workspace = torch.empty(
+            self.workspace_size, dtype=torch.uint8, device="cuda"
+        )
```
As we're saving the `workspace` as a member now, would it make sense to skip workspace size detection and initialization if it's already done?

jansel commented on torch/_inductor/utils.py:
```diff
@@ -45,6 +45,79 @@
 VarRanges = Dict[sympy.Expr, sympy.Expr]
 
 
+def do_bench_using_profiling(fn: Callable[[], Any], warmup=25, rep=100) -> float:
```
Why can't the existing do_bench be used?  This seems pretty similar.

ipiszy replied:
do_bench also measures CPU-side overhead.

e.g. for an event sequence:
CUDA event begin -> CUDA launch kernel -> CUDA kernel execution -> CUDA event end
It measures time between CUDA event begin and CUDA event end, which contains "CUDA launch kernel" part. This part could take some CPU time. This is especially bad for CUTLASS kernels which rely on ctypes to invoke C++ functions from Python.

do_bench_using_profiling, on the other hand, relies on profiler to collect kernel device time.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @jansel @aakhundov @kadeng , ptal~

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

aakhundov commented on test/inductor/test_inductor_utils.py:
```diff
@@ -0,0 +1,34 @@
+
+    def test_do_bench(self):
+        res = do_bench(self._bench_fn)
+        log.error("do_bench result: %s", res)
```
Maybe use `log.warning` instead?

jansel commented on test/inductor/test_inductor_utils.py:
```diff
@@ -0,0 +1,34 @@
+
+    def test_do_bench_using_profiling(self):
+        res = do_bench_using_profiling(self._bench_fn)
+        log.error("do_bench_using_profiling result: %s", res)
```
assert something about the return value

jansel commented on test/inductor/test_inductor_utils.py:
```diff
@@ -0,0 +1,34 @@
+        cls._bench_fn = functools.partial(torch.nn.functional.linear, x, w)
+
+    def test_do_bench(self):
+        res = do_bench(self._bench_fn)
```
tests should assert something

ipiszy replied:
There is a check to throw exception inside `do_bench_using_profiling()`, so the test below makes sure that no exception is thrown. I also use this two tests to compare latency collected by these two functions. We could skip the `test_do_bench()` test in CI though...

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
@pytorchbot label "topic: not user facing"

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest"

This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]
